### PR TITLE
armQw1ue: VSP to validate Level Of Assurance

### PR DESCRIPTION
**Pull request by AmritSidhu**
https://trello.com/c/armQw1ue/246-vsp-to-validate-level-of-assurance
VSP now validates against Level of Assurance using the LevelOfAssuranceValidator
Added unit tests to prove the correct exceptions are being thrown with the new validation
Added another acceptance test to make sure the correct exception is being thrown if loa returned by idp is too low

### CODE REVIEW

File: src/main/java/uk/gov/ida/verifyserviceprovider/services/NonMatchingAssertionService.java [hunk 1]
```diff
@@ -1,12 +1,9 @@
 package uk.gov.ida.verifyserviceprovider.services;
 
 import org.opensaml.saml.common.SAMLVersion;
-import org.opensaml.saml.saml2.core.Assertion;
-import org.opensaml.saml.saml2.core.AuthnStatement;
-import org.opensaml.saml.saml2.core.StatusCode;
+import org.opensaml.saml.saml2.core.*;
```
Comment (andy-paine) [inline src/main/java/uk/gov/ida/verifyserviceprovider/services/NonMatchingAssertionService.java]: Can you set your IDE to not use "wildcard imports"

Setting in IntelliJ is Preferences -> Editor -> Code Style -> Java -> Imports and set the two settings to 9999

File: src/main/java/uk/gov/ida/verifyserviceprovider/services/NonMatchingAssertionService.java [hunk 2]
```diff
@@ -64,7 +62,11 @@ public TranslatedNonMatchingResponseBody translateSuccessResponse(List<Assertion
         validate(authnAssertion, mdsAssertion, expectedInResponseTo, expectedLevelOfAssurance);
 
         String nameID = mdsAssertion.getSubject().getNameID().getValue();
+
         LevelOfAssurance levelOfAssurance = extractLevelOfAssuranceFrom(authnAssertion);
+        LevelOfAssuranceValidator levelOfAssuranceValidator = new LevelOfAssuranceValidator();
```
Comment (andy-paine) [inline src/main/java/uk/gov/ida/verifyserviceprovider/services/NonMatchingAssertionService.java]: In general we don't `new` our validators in the class as it makes it harder to unit test - dependencies should come in via the constructor (and thus the factory)

File: src/test/java/unit/uk/gov/ida/verifyserviceprovider/services/NonMatchingAssertionServiceTest.java [hunk 3]
```diff
@@ -34,11 +35,7 @@
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.util.Lists.emptyList;
 import static org.mockito.ArgumentMatchers.any;
-import static org.mockito.Mockito.doNothing;
-import static org.mockito.Mockito.mock;
-import static org.mockito.Mockito.times;
-import static org.mockito.Mockito.verify;
-import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.*;
```
Comment (andy-paine) [inline src/test/java/unit/uk/gov/ida/verifyserviceprovider/services/NonMatchingAssertionServiceTest.java]: Same as above

File: src/test/java/unit/uk/gov/ida/verifyserviceprovider/services/NonMatchingAssertionServiceTest.java [hunk 4]
```diff
@@ -204,6 +201,28 @@ public void shouldCorrectlyExtractLevelOfAssurance() {
...
 
 
+    @Test
+    public void shouldThrowExceptionWhenLevelOfAssuranceNotPresent() {
+        Assertion authnAssertion = anAuthnStatementAssertion(null, "requestId").buildUnencrypted();
+        Assertion mdsAssertion = aMatchingDatasetAssertionWithSignature(emptyList(), anIdpSignature(), "requestId").buildUnencrypted();
+
+        exception.expect(SamlResponseValidationException.class);
+        exception.expectMessage("Expected a level of assurance.");
+        nonMatchingAssertionService.translateSuccessResponse(ImmutableList.of(authnAssertion, mdsAssertion), "requestId", LEVEL_2, "default-entity-id");
```
Comment (andy-paine) [inline src/test/java/unit/uk/gov/ida/verifyserviceprovider/services/NonMatchingAssertionServiceTest.java]: If we provide the `LevelOfAssuranceValidator` in the constructor, we can provide a mock and `verify` on that mock here.

File: src/main/java/uk/gov/ida/verifyserviceprovider/services/NonMatchingAssertionService.java (hunk 1 above)
Comment (AmritSidhu) [reply src/main/java/uk/gov/ida/verifyserviceprovider/services/NonMatchingAssertionService.java]: Fixed imports👍🏼

File: src/main/java/uk/gov/ida/verifyserviceprovider/services/NonMatchingAssertionService.java (hunk 2 above)
Comment (AmritSidhu) [reply src/main/java/uk/gov/ida/verifyserviceprovider/services/NonMatchingAssertionService.java]: Done👍🏼

File: src/test/java/unit/uk/gov/ida/verifyserviceprovider/services/NonMatchingAssertionServiceTest.java (hunk 3 above)
Comment (AmritSidhu) [reply src/test/java/unit/uk/gov/ida/verifyserviceprovider/services/NonMatchingAssertionServiceTest.java]: Fixed imports👍🏼

File: src/test/java/unit/uk/gov/ida/verifyserviceprovider/services/NonMatchingAssertionServiceTest.java (hunk 4 above)
Comment (AmritSidhu) [reply src/test/java/unit/uk/gov/ida/verifyserviceprovider/services/NonMatchingAssertionServiceTest.java]: Implemented the mock use

**Comment by AmritSidhu** [comment]
@alan-gds isn't that what the test 'shouldThrowExceptionIfLoAReturnedByIdpIsTooLow' is doing?

**Comment by alan-gds** [comment]
D'oh, you're absolutely right; I was looking in the wrong file!  In that case, I'll approve this!

**Pull request merged**